Implement plaintext fallback when no keyring is available for MSAL cache persistence

Binary-Eater opened the pull request:
Users on headless systems may not have a supported keyring like a Secret
Service API provider. In those situations, fall back to using plaintext
storage of the MSAL cache. Throw a warning to indicate to the user that
this action occurred. The msal-python documentation does plaintext file
system storage for its cache persistence example.

Binary-Eater committed:
fixup! Isolate keyring instances based on tenant-id and client-id

Binary-Eater committed:
Support python standard metadata for packaging git-credential-msal

Will help simplify distribution-specific packaging work.

Link: https://peps.python.org/pep-0517/
Signed-off-by: Rahul Rameshbabu <sergeantsagara@protonmail.com>

Binary-Eater committed:
Convert README from orgmode to markdown

Markdown has better support across a wide range of editors. Orgmode
realistically only has first-class support in emacs.

Signed-off-by: Rahul Rameshbabu <sergeantsagara@protonmail.com>

Binary-Eater committed:
Implement insecure plaintext fallback when no keyring is available for MSAL cache persistence

Users on headless systems may not have a supported keyring like a Secret
Service API provider. In those situations, fall back to using plaintext
storage of the MSAL cache if the required flag is provided.The msal-python
documentation does plaintext file system storage for its cache persistence
example.

Link: https://msal-python.readthedocs.io/en/latest/#msal.SerializableTokenCache
Signed-off-by: Rahul Rameshbabu <sergeantsagara@protonmail.com>

Binary-Eater commented:
I originally wanted to use `pass` for this, so the fallback is GPG-encrypted.

https://github.com/Binary-Eater/git-credential-msal/commit/4f2ebbc0cbf9c1c9547d8893d284519bdecb49ea

The problem is that the modern 3rd party pass python keyring package does not invoke pass's multiline option, so it cannot store the serialized cache correctly.

https://github.com/nazarewk/keyring_pass/issues/13#issuecomment-1621595856

Not really interested in submitting a PR to this project and making it a dependency, even as an optional one.

aaronp24 commented on pyproject.toml:
```diff
@@ -0,0 +1,43 @@
+[tool.setuptools_scm]
+
+[tool.isort]
+profile = "black"
```
Should there be a newline at the end of this file?

Binary-Eater replied:
Yeah, there should be. Normally my editor automatically adds one. Weird. Should be fixed now.

aaronp24 commented on src/git_credential_msal/__init__.py:
I'm not too familiar with this Python packaging stuff so I'll take your word for it. It seems like it adds a lot of complexity, though.

Binary-Eater replied:
I am not either. I stumbled across this when Arch Linux recommended that python packages be written like this for the sake of the PKGBUILD.

* https://wiki.archlinux.org/title/Python_package_guidelines#Standards_based_(PEP_517)
* https://peps.python.org/pep-0517/

This change made packaging for Arch and NixOS pretty simple. As for ad-hoc testing, I found the compile from source instructions fairly clean and contained under `$HOME/.local` or you can directly run `main.py` with the python interpreter just like before this change for local testing.

Speaking of which, I discovered just how messy the python ecosystem is. There are numerous package managers.

* [setuptools](https://setuptools.pypa.io/en/latest/) + [setuptools_scm](https://setuptools-scm.readthedocs.io/en/latest/)
* [pdm](https://pdm-project.org/en/latest/)
* [poetry](https://python-poetry.org/)
* .....

Basically, PEP 517 introduced an abstraction interface over the :infinity: package managers in Python.....

https://drewdevault.com/2021/11/16/Python-stop-screwing-distros-over.html

It seems implementing PEP 517 is the best way to simplify the work in distribution packaging. It also offers me magic versioning functionality, which I think is helpful. That said, I understand your point about complexity introduced.

aaronp24 replied:
Thanks for the Arch link. I'll try updating my PKGBUILD to use that and see how it goes.

aaronp24 commented on src/git_credential_msal/main.py:
```diff
@@ -147,11 +161,23 @@ def get_msal_cache(name: str) -> SerializableTokenCache:
+            msal_cache_name = f"msal_cache_{name}"
+            msal_cache_path = os.path.join(cache_dir, msal_cache_name)
+            with open(msal_cache_path, "w") as f:
+                f.write(data)
```
This will automatically write secrets to the disk unencrypted with no user intervention, right? I'm not a huge fan of that idea. Would it make sense to print an error instead and refuse to write the secrets unless users pass an additional flag on the command line, or maybe set an additional git config flag? That could also silence the warning if users are explicitly opting in.

Binary-Eater replied:
I also was not a huge fan of this idea. However, I wanted to come up with a solution that would satisfy the use case of headless machine users. I originally authored an encrypted solution but I wanted to minimize introducing more external dependencies.

https://github.com/Binary-Eater/git-credential-msal/pull/9#issuecomment-2182322510

I like your suggestion. Will go forward with it.

Binary-Eater replied:
I feel like I came up with a better solution security-wise by depending on other credential helpers (especially the git-credential-cache helper). Let me know what you think.

Binary-Eater replied:
No refresh token caching in this model, but I think that's fair in an ecosystem where the user chooses not the use a keyring provider. Using the cache helper makes it so such a use case only involves one auth dance in a partial clone operation in theory.

aaronp24 replied:
Right, but isn't any later use going to require another auth dance? That seems pretty user unfriendly.

Given that git-credential-store doesn't seem well suited to this, it does seem worth having git-credential-msal be able to store the refresh token, with an opt-in if it's insecure.

Binary-Eater replied:
https://github.com/Binary-Eater/git-credential-msal/pull/9/commits/1046d3d623c09587afb894fd4c0e0614672c7623

Binary-Eater replied:
I had some iterations from testing :joy:, https://github.com/Binary-Eater/git-credential-msal/commit/e74b662c84e516a8f5159537428299720fd23038

Binary-Eater committed:
Do not raise an exception if no keyring provider is present

The credential helper is still functional without a keyring provider. The
use of auxiliary credential helpers makes it viable during a working
session.

Signed-off-by: Rahul Rameshbabu <sergeantsagara@protonmail.com>

aaronp24 commented:
It might still be nice to allow users to save the refresh token in plaintext, though, if they want to. I wonder if it's possible to pass multiple `store` commands back to the caller and use that to store both tokens in a different credential helper. I don't see a git-credential-keyring but maybe that would be worth creating, and the existing git-credential-store works for storing unencrypted data.

The git-credential protocol does have a `oauth_refresh_token` field but it seems pretty half-baked and I'm not sure it's usable for this.

aaronp24 commented:
Never mind, that doesn't make sense: git-credential itself is what invokes the `store` command, and it does it with a single credential that it's approving. git-credential-msal would have to invoke its own copy of `git credential approve`.

Binary-Eater commented:
A couple of points (though I think you have discussed them yourself).

>It might still be nice to allow users to save the refresh token in plaintext, though, if they want to.

I have code for this that I stashed away. What I will do is leave the present behavior as is (with no warning prints, but the flag I have will enable insecure persistence of the MSAL cache).

>I wonder if it's possible to pass multiple store commands back to the caller and use that to store both tokens in a different credential helper.

I think the first issue is that unlike the `cache` credential helper, the `store` credential helper depends on the git uri parsing and serialization for persisting credentials. There is no uri information component where the cache value could be tacked on.

This leads into the second point that you touched open. We are not submitting the cache (`oauth_refresh_token`) to the server, so I assume the invocation for storing the credentials would not be able to cache this either way.

>I don't see a git-credential-keyring but maybe that would be worth creating, and the existing git-credential-store works for storing unencrypted data.

Similar issue as previous point.

>The git-credential protocol does have a oauth_refresh_token field but it seems pretty half-baked and I'm not sure it's usable for this.

Yeah, I was thinking about this too. My issue is the MSAL cache is formatted in a way controlled by the MSAL library. I didn't want to extract pieces and try rebuilding the format in case Microsoft decides to break how they handle the cache data in future iterations.

aaronp24 commented on src/git_credential_msal/main.py:
```diff
@@ -134,31 +134,56 @@ def extract_entra_ids_from_git_config(helper_pairs: dict[str, str]) -> tuple[str
         data = keyring.get_password("system", name)
     except keyring.errors.NoKeyringError:
-        pass
+        data = get_msal_cache_insecure(name, allow_insecure)
```
Would it make sense to move this out of the `except` block and just do
```
if not data and allow_insecure:
    data = get_msal_cache_insecure(name)
```
That way, if someone stores the msal cache insecurely and then later logs into a desktop environment that has Keyring but it doesn't have a stored msal token, it can still find the insecure one?

Binary-Eater replied:
I was thinking about this too and wasn't sure what would be the better design choice. Will support this model. However, if someone installs a keyring, they may not invoke the `--insecure` flag. Would it make sense to still read from an already written insecure cache entry in that situation? Basically omit the `and allow_insecure`.

aaronp24 replied:
I think it would make sense to allow reading an insecure credential even without `--insecure`. Worse-case it just isn't valid, right? I guess it might be weird to have it read an insecure credential and then later write it to Keyring, leaving stale credentials in the cache. But as long as stale credentials are silently invalidated if they don't work it's probably fine?

Binary-Eater replied:
Yeah, that's exactly the line of thinking that went through my mind as well.

aaronp24 commented on README.md:
```diff
@@ -40,6 +40,11 @@ keyring provider present, so you can use the auxiliary credential helper to
 
+**NOTE:** This does not store the cache related to the OAuth2.0 refresh token
+that is stored in the keyring. This cache is useful for not needing to
+re-authenticate for 90-days. Have a look at the Device Authorization Grant
```
Is it always 90 days or is that configurable in the Azure tenant? `s/90-days/90 days/` otherwise.

Binary-Eater replied:
It's not configurable by the Azure tenant or tenant owner. However, the expiration could be sooner if certain account actions occur to invalidate it sooner.

* https://learn.microsoft.com/en-us/entra/identity-platform/refresh-tokens#token-lifetime
* https://learn.microsoft.com/en-us/entra/identity-platform/refresh-tokens#token-revocation

Binary-Eater replied:
I should probably provide these links in the README........

aaronp24 replied:
Thanks, that makes sense. Maybe just "for up to 90 days" then?

Binary-Eater replied:
Done :). I included the links as well for any user that might be interested.
